### DISCUSSION TITLE
Why are some permissionSources empty?

### DISCUSSION THREAD

**Discussion by fei0x** [community/community#24829] [Apps, API and Webhooks]
<p>I am using the graphql api to query RepositoryCollaboratorEdge objects to see who has access to my repo and how.</p>
<p>But I am noticing that some users with permissions have nothing in the permissionSources. (To be clear, I am seeing organizations, teams and repostiory objects returning in the same query for other repos/collaborators.)</p>
<p>For example. This query:</p>
<pre><code>query {
	viewer {
		repositories ( first:1 affiliations:[OWNER, ORGANIZATION_MEMBER], ownerAffiliations:[OWNER, ORGANIZATION_MEMBER] ){
			totalCount
			pageInfo {
				endCursor
				hasNextPage
			}
			nodes {
				name
				nameWithOwner

				collaborators (	first:1 ) {
					totalCount
					pageInfo {
						endCursor
						hasNextPage
					}
					edges {
						node { login }
						permission
						permissionSources {
							permission
							source {
								... on Organization { websiteUrl id name }
								... on Repository { homepageUrl id name nameWithOwner }
								... on Team { editTeamUrl id name combinedSlug }
							}
						}
					}
				}
			}
		}
	}
}
</code></pre>
<p>Has some results for collaborators that are looking like this snippet:</p>
<pre><code>"collaborators": {
                            "totalCount": 103,
                            "pageInfo": {
                                "endCursor": "Y3Vyc29yOnYyOpHNA8E=",
                                "hasNextPage": true
                            },
                            "edges": [
                                {
                                    "permission": "WRITE",
                                    "permissionSources": []
                                },
                                {
                                    "permission": "ADMIN",
                                    "permissionSources": []
                                },
</code></pre>
<p>So my questions is: <strong>What might cause these permission sources arrays to be empty?</strong></p>
<p>Maybe this is an issue with the permission levels of the caller? But it seems strange to me that I should be able to see the collaborators and there permission at all if i can’t see the permissionSources.  Or perhaps there is another way to become a collaborator beyond organization membership, team membership or direct repo assignment? I’m also noticing that this appears to be happening to repos as a whole so maybe there is something there as well?</p>

**Comment by AndreaGriffiths11** [comment, marked as answer]
<p>Hi <a class="mention" href="/fei0x">@fei0x</a>,</p>
<p>Thanks for reaching out! If the token you are using doesn’t have the correct (<code>admin:org</code>) scope, the GraphQL API will return a message saying so. If the token has this scope, but the user isn’t an admin on the organization, nothing is returned, only a blank array.</p>
<p>I suspect that this information requires a user to be an organization administrator because this is the only way to find out this information in the web interface as well. Having this information available to every user through the API may be deemed a security risk.</p>
<p>I haven’t done exhaustive testing, but it looks like a GitHub App will need <a href="https://developer.github.com/v3/apps/permissions/#metadata-permissions" rel="nofollow noopener">metadata permission</a> to find out information on collaborators. This gives the same information as the <code>permissionSources</code> endpoint.</p>
<p>I hope this helps!</p>

**Reply by toniopelo** [reply]
Hi @AndreaGriffiths11,
I can't find any way to get the `permissionSources` as a Github App, the return is always an empty array.
I even granted my test Github App all the possible permissions there is, and still I get an empty array.
How can I achieve that ? I tried with a personal access token generated from my user settings and it works, so the Graphql request is fine.
Thanks for your time!

**Reply by yuvalyacoby** [reply]
for me setting the Organization -> Admin read only solved it

**Comment by haridsv** [comment]
I am an admin and I have given `admin:org > read:org` permission, but Istill get an error that the `admin:org` scope is missing. The query goes through if I give full `admin:org` scope (i.e., including `write:org`), but why does a read-only query need write scope?

NOTE: giving `read:org` and `write:orrg` scopes individually is not enough, you need to give the entire `admin:org` scope (though there are no other sub-scopes), which sort of indicates a bug in the scope check logic.
